docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+```
+
+would be transformed into those fields:
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/GO.jpg"
+}
+```
+
+This parser use [glob pattern matching](https://en.wikipedia.org/wiki/Glob_(programming)) and extract named matches(*\<name>) into fields.
+If the line doesn't match the glob it will be filtered out.
+
+### Json
+
+The json parser allows to parse a json log line and select fields to extract. If the line is not a valid json document the line will be skipped.
+
+We could support those notations:
+
+```logql
+{job="prod/app"} | json .route client_ip=.client.ip
+{job="prod/app"} | json .route .client.ip as route, client_ip
+{job="prod/app"} | json |select route client_ip
+{job="prod/app"} | json
+```
+
+> `{job="prod/app"} |json |select route client_ip` is an implicit extraction, the list of extracted field is inferred from the full query.
+
+The log line :
+
+```log
+{
+    "route": "/api/v1/foo",
+    "client": {
+        "ip": "127.0.0.1",
+        "org-id": "grafana-labs"
+    }
+}
+```
+
+would be transformed into those fields:
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "route": "/api/v1/foo"
+}
+```
+
+### Logfmt
+
+The logfmt parser will allow to extract fields from the [log format](https://brandur.org/logfmt). For example the expression:
+
+```logql
+{job="prod/app"} | logfmt client_ip path
+```
+
+will transform the log line
+
+```log
+time=03-10-2020T20:00:01 path=/foo/bar client_ip="127.0.0.1"
+```
+
+into those fields:
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/foo/bar"
+}
+```
+
+Again it might be useful to allow renaming fields to avoid conflicts with labels.
+
+For instance `{job="prod/app"} | logfmt ip=client_ip path` will rename the field `client_ip` into `ip`.
+
+`logfmt` will also support implicit field extraction like `json`.
+
+### Regexp
+
+At last, the regexp operator should cover all other use cases but is probably the less efficient and intuitive.
+
+> While we already support `|~` and `!~` filters, we don't want to mix parsers and filters since their purpose are different, filters should be used to narrow the log volume and reduce the number of lines to parse. This is why we will use the notation `| regexp` for the parser operator.
+
+For example the expression `{job="prod/app"} | regexp "(?<client_ip>.+) - - (?<path>.+)"` will transform the log line `127.0.0.1 - - /foo/bar` into :
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/foo/bar"
+}
+```
+
+### Questions
+
+- Should we allow to implicit field extraction such as `{job="prod/app"} | logfmt` and `{job="prod/app"} | json` would extract all fields ? I really like the simplicity of it.
+- In the previous document the `as field_name1, field_name2` field renaming notation was also interesting should we use is over `ip=client_ip` for now I've opted for the later which can be omitted in case of no renaming ?
+- Do we prefer `|json` or `| json` notation ?
```

Comment:
Why should we choose one? Can't both be supported? If not, I would pick `| json` because `|something` is currently used for the filter expression (ie. `|=`) while here we're defining subsequent processing stages.

Reply:
Yep we can I thin that's would be likely.